lib/catalog/workspace_builder.rb

```diff
@@ -0,0 +1,37 @@
+module Catalog
+  class WorkspaceBuilder
+    attr_reader :workspace
+
+    def initialize(order)
+      @order = order
+    end
+
+    def process
+      @workspace = {'user' => user, 'request' => request, 'artifacts' => artifacts}
+
+      self
+    end
+
+    private
+
+    def user
+      usr = Insights::API::Common::Request.current.user
+      {'email' => usr.email, 'name' => "#{usr.first_name} #{usr.last_name}"}
+    end
+
+    def request
+      # assume there is only one applicable
+      applicable = @order.order_items.find_by(:process_scope => 'applicable')
+      {'order_id' => @order.id, 'order_started' => @order.order_request_sent_at, 'order_params' => applicable.service_parameters_raw}
+    end
+
+    def artifacts
+      facts = {'before' => {}, 'applicable' => {}, 'after' => {}}
```

Comment:
@bzwei @eclarizio Should applicable be called current meaning the Curren order. When you look at it from a substitution string pattern what looks better
```
{{applicable.order_params.param1}} or
{{current.order_params.param1}}
```

Reply:
I am ok with either name. These names are allowed values for the `process_scope` attribute. I think they are good from coding point of view.

Reply:
Prefer using the same word in artifact key and process_scope attribute. May I suggest `main` the word to replace `applicable`?

Reply:
@bzwei I think We are missing the params from the pre and post. I am wondering if the hash for pre, post and applicable have to follow a pattern

```
{ params => {}, artifacts => {}, status => {} }
```
This would allow them to do substitution from any of the pre/post/applicable in a consistent format.

If they have a hidden param in the pre and they wanted to use it in post they might need access to it. Otherwise everything would have to come thru artifacts even the input parameters. We have to think of params as input and artifacts as output.

Reply:
@mkanoor I was following the pattern in your initial email. Can you update the (complete) pattern with your new proposal in Jira?

Reply:
@mkanoor @bzwei I could go either way on `applicable` vs. `current`, but I think in either case, if we ever support a "cart" type system with multiple products in it (as in, multiple "applicables"), I think `applicable` makes more sense, because the order process knows which product it is applied to, and if there are multiple products, it may not be what is necessarily being handled currently. With multiple products, `main` might not fit as well, because it gives the sense that for other products you might use a `secondary`, `tertiary`, `fourth`(?), instead of just another `main`.

For how we do things now, though, `current` would be fine if we want to change to that. I don't think it's difficult to change, though I suppose we probably want to set a precedent if we're telling users what to use for their substitution values, and changing it later down the road might lead to doc changes and customer training that has more of an impact.

Obviously, I'm biased to `applicable` since I spent quite a bit of time (probably too long, hah) coming up with it, but if you guys don't necessarily agree with my argument for it, we can use `current` instead. 👍

Reply:
@bzwei @eclarizio 
I think this might be a better layout it allows us to pick things up from an order item in a consistent fashion for pre/post and the main/applicable/current

each order item has the 
     

1. params
2. artifacts
3. status

Picked up directly from the order item

```
{
  'user' => {"email" => "joe@google.com","name" => "Joe DiMaggio"},
  'request' => {'order_id' => 123, "order_started" => "20201001T11:11:11Z"},
  'order_process' => {
    'before' => {
      'opname1' => { 'artifacts' => {'ip' => "1.1.1.94", 'ticket' => "abc23455", 'priority' => "high"},
                                'params'   => { 'var1' => 2, 'var2' => 'low'},
                                'status'   => "successful"
                   }
      'opname2' => { 'artifacts' => {'caller' => "catalog", 'level'  => 'debug'},
                                'params'   => { 'name' => "Fred", 'age' => 52},
                                'status'    => "successful"
                   }
    },
    'after' => {
      'opname1' => {'artifacts' => { 'closed_at' => '2020', 'status'    => 'complete'},
                                'params'    => {},
                                'status'    => 'failed'
                   },
      'opname2' => {'artifacts' => { 'support_email' => 'support@slate.com', 'support_tech'  => 'Fred Flintstone'},
                                'params'    => {},
                                'status'    => 'successful'
                   }
    }
  }
  'current' => { 'artifacts' => { 'var2' => '2', 'var3'  => '3'},
                         'params'    => {},
                         'status'    => 'successful'
  }
}
```

I am not sure if we need the order_process node in between we could just call it before/after/current and drop the order process. With this approach we have access to the input (params) output(artifacts) and the status of each of the order items.

Reply:
Updated with above template, with slight modification: the applicable item has the same pattern as the before/after item. This will allow to expand to multiple applicable in the future.
```
{
  'user' => {"email" => "joe@google.com","name" => "Joe DiMaggio"},
  'request' => {'order_id' => 123, "order_started" => "20201001T11:11:11Z"},
  'before' => {
      'opname1' => { 'artifacts' => {'ip' => "1.1.1.94", 'ticket' => "abc23455", 'priority' => "high"},
                                'extra_vars'   => { 'var1' => 2, 'var2' => 'low'},
                                'status'   => "successful"
                   }
      'opname2' => { 'artifacts' => {'caller' => "catalog", 'level'  => 'debug'},
                                'extra_vars'   => { 'name' => "Fred", 'age' => 52},
                                'status'    => "successful"
                   }
  },
  'after' => {
      'opname1' => {'artifacts' => { 'closed_at' => '2020', 'status'    => 'complete'},
                                'extra_vars'    => {},
                                'status'    => 'failed'
                   },
      'opname2' => {'artifacts' => { 'support_email' => 'support@slate.com', 'support_tech'  => 'Fred Flintstone'},
                                'extra_vars'    => {},
                                'status'    => 'successful'
                   }
  }
  'applicable' => {
      'appname' => { 'artifacts' => { 'var2' => '2', 'var3'  => '3'},
                         'extra_vars'    => {},
                         'status'    => 'successful'
      }
  }
}
```

Reply:
👍

Reply:
@bzwei The app name under applicable is a problem because it can have embedded spaces and characters that will cause mustache issues. Unlike the OrderProcess we are not enforcing what characters are allowed in the name of a PortfolioItem

Reply:
@mkanoor I think this is the story for that https://issues.redhat.com/browse/SSP-1870